fix: Connecting to Datasource - Rehaul (#691)
appsmithorg/appsmith-docs: website/docs/core-concepts/connecting-to-data-sources/authentication/connect-to-apis.md

# Configuring APIs

Configuring APIs involves setting up and managing the various aspects of an API that allow it to function effectively. This can include defining the API's structure and capability, setting up authentication and authorization, and establishing connections with other systems or services.

:::tip
If you want to access an API that's not publicly available on the internet, you can use a service like [ngrok](https://ngrok.com/) to expose it. For more information, see [How to work with local APIs on Appsmith](/advanced-concepts/more/how-to-work-with-local-apis-on-appsmith).
:::

> ## API query editor

Reviewer: Can we replace `API query editor` with `REST API query editor` - since API can mean any protocol (non REST) as well. Similar for other usage below as well.

Reviewer 2: FYI: Changed this back to API pane as in the current docs. not sure if we change it to query editor.

> The API pane is a REST interface that allows you to create and modify your existing APIs. All [REST HTTP methods](https://www.w3schools.in/restful-web-services/rest-methods) are supported and API values can be configured in the headers, Params, and body fields

Reviewer: Sorry I could not understand this line at all - `The API pane is a REST interface that allows you to create and modify your existing APIs.`

Author: rephrased

> The API pane is a REST interface that allows you to create and modify your existing APIs. All [REST HTTP methods](https://www.w3schools.in/restful-web-services/rest-methods) are supported and API values can be configured in the headers, Params, and body fields

Reviewer: Sorry I could not understand the meaning of API values in this line: `supported and API values ...`

Author: that you can configure different things from the API pane

![](</img/create_api_(1).gif>)

## Configure API query

> Configuring an API query involves specifying the various components that make up the request and the desired response. Here is an overview of the main elements you need to consider when configuring an API query.

Reviewer: Can we replace API with REST API ?

Author: done

Reviewer 2: @sumitsum Hey do we need to mention REST everywhere? In the beginning and in the intro wouldn't it suffice to add REST and then avoid it in sub-headings?

### Name

> When you create a datasource in Appsmith, a default name is provided for you. You can change this name by clicking on the pencil icon. It's important to make sure that the name is unique, as it serves as an identifier for queries on the page and is used to access the datasource properties in the application.

Reviewer: The first line contains the work `datasource`. Should it be `query` ?

Author: dome

> When you create a datasource in Appsmith, a default name is provided for you. You can change this name by clicking on the pencil icon. It's important to make sure that the name is unique, as it serves as an identifier for queries on the page and is used to access the datasource properties in the application.

Reviewer: Sorry I could not understand this part - `is used to access the datasource properties in the application.`

Author: updated

### Method

> A method refers to the type of request being made to the API. Appsmith supports all the standard methods, including GET, POST, PUT, DELETE, and PATCH. When setting up an API query in Appsmith, you can select the desired method from a dropdown menu before adding the URL.

Reviewer: Please replace `request being made to the API` with `request being made to the REST API endpoint`.

It's important to choose the appropriate method based on the task you want to accomplish, as each method has a specific meaning and can behave differently depending on the API you are using.

### URL

The URL Path provides a way to access the API and its functions. To use this field, you can enter the API URL that you want to access. For example, if you want to access the Appsmith mock API, you can enter the URL as `https://mock-api.appsmith.com`.

### Headers

HTTP headers are a critical part of the API request and response, as they contain meta-data about the API request and response. Headers carry the following information:

> * Request and Response Body;

Reviewer: Is the request and response body carried as part of the headers ? - may not be correct. Can you please check this once ?

Author: I guess the request and response body are not carried as part of the headers. The headers contain metadata about the request or response. REMOVED that  part

Author: @sumitsum can you confirm this once?

> * Request and Response Body;

Reviewer: Can we remove this list - because it does not seem exhaustive i.e. covers very few things, also it would not be too long to list down everything that header can include.

Author: removed

* Request Authorization;
* Response Caching;
* Response Cookies.

It's important to be familiar with the major request and response headers, as you may need to set request headers when testing an API, and you may need to set assertions against response headers to ensure that the correct headers are being returned.

> ### Query parameters

Reviewer 2: This is **Params** in API configuration

> In a REST API, parameters are values that are passed in the URL, query string, or request body that are used to filter, sort, paginate, or perform some other operation on a set of data. These parameters allow the API to return a specific set of data based on the values that are provided.

Reviewer: Sorry, can you please check this once, IMO, params are always passed as part of the URL only.

Author: updated

> In a REST API, parameters are values that are passed in the URL, query string, or request body that are used to filter, sort, paginate, or perform some other operation on a set of data. These parameters allow the API to return a specific set of data based on the values that are provided.

Reviewer: `that are used to filter, sort, paginate, or perform some other operation on a set of data` - this is not a exhaustive list - params could be used for other things as well. Can we re-write this line ?

Author: @sumitsum can you suggest a few things?

Reviewer: @harshilp24 IMO, it should be good if we just change the following line `that are used to filter` to `that can be used to filter`.

Author: done

### Body

In an API query, the body is the main content of the request. It's typically used to send data to the server, such as form data or a JSON payload. The body of an API request is usually included with POST and PUT requests, which are used to create or update resources on the server. For example, it might contain a JSON object with multiple key-value pairs, or it might contain a simple string or number.

### Pagination

Pagination refers to the process of receiving portions of a huge dataset until you receive the entire dataset. This helps optimize your app's performance because trying to get all the data results in a slow load time of your app and a slow rendering of the results in the UI. Appsmith supports the following methods for paginating API queries:

#### Paginate with table page number
Allows you to paginate the results of an API query using a table page number. To use this feature, you need to configure table and request parameters.

This can be done by setting up request parameters to control the table's pagination. This may involve mapping a key like "**pageNo**" to the table's page number property.

> Example - Map key pageNo or similar to value

Reviewer: While this may work out to be correct in certain cases, we should also provide one more example where we may want to configure `limit` and `offset` property separately. e.g. `limit = {{UsersTable.pageSize}}` and `offset={{UsersTable.pageOffset}}`

Author: done

```{{UsersTable.pageNo}}``` 

#### Paginate with response URL

> A response URL is a special type of URL that's returned in the API response and can be used to request the next or previous page of results. To paginate with response URLs, the API typically includes a link to the next or previous page of results in the HTTP header of the response. This can be done by adding **Previous** and **Next URL** in the Pagination section. 

Reviewer: `includes a link to the next or previous page of results in the HTTP header of the response.` - this does not seem right. Currently we expect the next or previous page URL to be part of the API response itself since we cannot access other API properties like response header etc.

Author: removed that line

> A response URL is a special type of URL that's returned in the API response and can be used to request the next or previous page of results. To paginate with response URLs, the API typically includes a link to the next or previous page of results in the HTTP header of the response. This can be done by adding **Previous** and **Next URL** in the Pagination section. 

Author: added

> A response URL is a special type of URL that's returned in the API response and can be used to request the next or previous page of results. To paginate with response URLs, the API typically includes a link to the next or previous page of results in the HTTP header of the response. This can be done by adding **Previous** and **Next URL** in the Pagination section. 

Author: done

# Configuring APIs

Configuring APIs involves setting up and managing the various aspects of an API that allow it to function effectively. This can include defining the API's structure and capability, setting up authentication and authorization, and establishing connections with other systems or services.

:::tip
If you want to access an API that's not publicly available on the internet, you can use a service like [ngrok](https://ngrok.com/) to expose it. For more information, see [How to work with local APIs on Appsmith](/advanced-concepts/more/how-to-work-with-local-apis-on-appsmith).
:::

## REST API query editor

> The API pane is a user-friendly interface for creating and managing RESTful APIs. It supports all standard [REST HTTP methods](https://www.w3schools.in/restful-web-services/rest-methods) and allows you to configure API values in the headers, query parameters, and request body fields.

Author: done

When you create a query in Appsmith, a default name is provided for you. You can change this name by clicking on the pencil icon. It's important to make sure that the name is unique, as it serves as an identifier for queries on the page.

### Method

A method refers to the type of request being made to the REST API endpoint. Appsmith supports all the standard methods, including GET, POST, PUT, DELETE, and PATCH. When setting up an API query in Appsmith, you can select the desired method from a dropdown menu before adding the URL.

It's important to choose the appropriate method based on the task you want to accomplish, as each method has a specific meaning and can behave differently depending on the API you are using.

### URL

The URL Path provides a way to access the API and its functions. To use this field, you can enter the API URL that you want to access. For example, if you want to access the Appsmith mock API, you can enter the URL as `https://mock-api.appsmith.com`.

### Headers

HTTP headers are a critical part of the API request and response, as they contain meta-data about the API request and response. Headers carry the following information:

* Request Authorization;

> * Response Caching;

Author: Done. removed the list.

* Response Cookies.

> It's important to be familiar with the major request and response headers, as you may need to set request headers when testing an API, and you may need to set assertions against response headers to ensure that the correct headers are being returned.

Reviewer: Can we remove this line - `and you may need to set assertions against response headers to ensure that the correct headers are being returned.` - reason being that at this moment, IMO, we don't provide a way to the user to check the response header values.

Author: done

When you want to ensure that the incoming requests originate from Appsmith, you can enable `Send Appsmith Signature Header` by selecting **Yes**. You’ll see a new field - **Session Details Signature Key** to supply the signature key.

![Appsmith Signature Header](</img/OAuth__API_Integration__Appsmith_Signature_Header__Enable__.png>)

### Authentication type

Common authentication types for REST APIs include: 

* [Basic Authentication](/core-concepts/connecting-to-data-sources/authentication/authentication-type/basic-authentication), where you have to provide a username and password. 
* [OAuth](/core-concepts/connecting-to-data-sources/authentication/authentication-type/oauth2-authentication), where you need to obtain a token from an authorization server and include that token to authenticated. 

The authentication type that's used by an API can have a significant impact on its security, so it's important to choose an appropriate authentication method for your API.

You can define an [authentication type](/core-concepts/connecting-to-data-sources/authentication/authentication-type) for REST APIs by using the protocols available on Appsmith.

### Settings

> The Query Settings Pane provides access to properties for the query. You can specify the following [settings](/core-concepts/data-access-and-binding/querying-a-database/query-settings) in the Settings tab on the Appsmith Query Editor.

Reviewer 3: ```suggestion
To access and configure settings for a query, click the Settings tab on the Appsmith Query Editor. Some options available in the Query Settings Pane include Run on Page Load and others. For more information, see [Query Settings](/core-concepts/data-access-and-binding/querying-a-database/query-settings).
```

# Configuring APIs

Configuring APIs involves setting up and managing the various aspects of an API that allow it to function effectively. This can include defining the API's structure and capability, setting up authentication and authorization, and establishing connections with other systems or services.

:::tip
If your API isn't available on the public internet, you must expose it via a service such as [https://ngrok.com/](/advanced-concepts/more/how-to-work-with-local-apis-on-appsmith#using-ngrok)
:::

## API query editor

The API pane is a REST interface that allows you to create and modify your existing APIs. All [REST HTTP methods](https://www.w3schools.in/restful-web-services/rest-methods) are supported and API values can be configured in the headers, Params, and body fields

![](</img/create_api_(1).gif>)

## Configure API query

Configuring an API query involves specifying the various components that make up the request and the desired response. Here is an overview of the main elements you need to consider when configuring an API query.

> | Property                       	| Configure                    	| Description                                                                                                                                                      	|

Reviewer 3: Let's rename the heading as 

| Setting | Availability | Description |

|--------------------------------	|------------------------------	|------------------------------------------------------------------------------------------------------------------------------------------------------------------	|

> | Name                           	| REST API & Authenticated API 	| The `name` property allows you to specify a name for the API.                                                                                                    	|

Reviewer 3: For the description column- let's follow like below:

| Name   | REST API & Authenticated API 	| Specify a name for the API |

Author: done

| Headers                        	| REST API & Authenticated API 	| Headers are key-value pairs that can be sent along with an HTTP request to provide additional information about the request context.                             	|
| Query parameters               	| REST API & Authenticated API 	| Query parameters are used to filter and sort the data that's being retrieved, or to specify certain options for the API's behavior.                             	|
| Body                           	| REST API                     	| The body is the part of the request that contains the payload of the request, which can include data that's sent to the server for processing or storage.       	|
| Pagination                     	| REST API                     	| Pagination is the process of dividing a large set of data into smaller chunks or pages, and providing a way for the client to access the data one page at a time 	|
| Send appsmith signature header 	| Authenticated API            	| When you want to ensure that the incoming requests originate from Appsmith, you can enable Send Appsmith Signature Header property.                              	|

### Name

> By default, Appsmith provides a default name for the data source it creates. You can edit this name by clicking on the pencil icon next to it. It's important to ensure that the name is unique, as it serves as an identifier for the queries on the page and is used to access the data source's properties throughout the application.

Reviewer 3: ```suggestion
When you create a datasource in Appsmith, a default name is provided for you. You can change this name by clicking on the pencil icon. It's important to make sure that the name is unique, as it serves as an identifier for queries on the page and is used to access the datasource properties in the application.
```

### Method

> An API query involves making a request to an API using a specific method, or type of request. Appsmith supports all the common methods, including GET, POST, PUT, DELETE, and PATCH. When configuring an API query in Appsmith, you can select the desired method from a dropdown menu before adding the URL. 

Reviewer 3: ```suggestion
A method refers to the type of request being made to the API. Appsmith supports all the standard methods, including GET, POST, PUT, DELETE, and PATCH. When setting up an API query in Appsmith, you can select the desired method from a dropdown menu before adding the URL.
```

# Configuring APIs

Configuring APIs involves setting up and managing the various aspects of an API that allow it to function effectively. This can include defining the API's structure and capability, setting up authentication and authorization, and establishing connections with other systems or services.

:::tip

> If your API isn't available on the public internet, you must expose it via a service such as [https://ngrok.com/](https://ngrok.com/)

Reviewer 3: The cross-reference should be to the local DB and API page for using ngrok.

Reviewer 3: This is still open. Please add reference to - /advanced-concepts/more/how-to-work-with-local-apis-on-appsmith#using-hostdockerinternal

:::

## API query editor

The API pane is a REST interface that allows you to create and modify your existing APIs. All [REST HTTP methods](https://www.w3schools.in/restful-web-services/rest-methods) are supported and API values can be configured in the headers, Params, and body fields

![](</img/create_api_(1).gif>)

> ## Configure API query

Reviewer 3: Too many repetitions here. Can we have the common properties listed and then, for specific properties add the type of APIs?
